mmengine/optim/scheduler/param_scheduler.py

```diff
@@ -1283,3 +1283,203 @@ def get_position_from_periods(
+            rule=rule, threshold=threshold, threshold_rule=threshold_rule)
+        self._reset()
+        self.step()
+
+    def load_state_dict(self, state_dict):
+        self.__dict__.update(state_dict)
+        self._init_is_better(
+            rule=self.rule,
+            threshold=self.threshold,
+            threshold_rule=self.threshold_rule)
+
+    def step(self):
+        self._global_step += 1
+
+        # Compute parameter value per param group in the effective range
+        if self.begin <= self._global_step < self.end:
+            # convert `metric` to float, in case it's a zero-dim Tensor
+            self.metric = float(self.metric)
+
+            self.last_step += 1
+
+            if self.is_better(self.metric, self.best):
+                self.best = self.metric
+                self.num_bad_epochs = 0
+            else:
+                self.num_bad_epochs += 1
+
+            if self.in_cooldown:
+                self.cooldown_counter -= 1
+                self.num_bad_epochs = 0  # ignore any bad epochs in cooldown
+
+            if self.num_bad_epochs > self.patience:
+                values = self._get_value()
+
+                for i, data in enumerate(
+                        zip(self.optimizer.param_groups, values)):
+                    param_group, value = data
+                    if param_group[self.param_name] - value > self.eps:
+                        param_group[self.param_name] = value
+                        self.print_value(self.verbose, i, value)
+                self.cooldown_counter = self.cooldown
+                self.num_bad_epochs = 0
+
+        self._last_value = [
+            group[self.param_name] for group in self.optimizer.param_groups
+        ]
+
+    def print_value(self, is_verbose: bool, group: int, value: float):
+        if is_verbose:
+            step_name = 'epoch' if self.by_epoch else 'iter'
+            print_log(
+                f'Adjusting parameter value of group {group} to {value:.4e} '
+                f'in {step_name} {self.last_step}.',
+                logger='current')
+
+    def _get_value(self):
+        values = [
+            float(group[self.param_name]) * self.factor
+            for group in self.optimizer.param_groups
+        ]
+        return [max(v, min_v) for v, min_v in zip(values, self.min_value)]
+
+    @property
+    def in_cooldown(self):
+        return self.cooldown_counter > 0
+
+    def is_better(self, a, best):
+        if self.rule == 'min' and self.threshold_rule == 'rel':
+            rel_epsilon = 1. - self.threshold
+            return a < best * rel_epsilon
+
+        elif self.rule == 'min' and self.threshold_rule == 'abs':
+            return a < best - self.threshold
+
+        elif self.rule == 'max' and self.threshold_rule == 'rel':
+            rel_epsilon = self.threshold + 1.
+            return a > best * rel_epsilon
+
+        else:  # rule == 'max' and epsilon_mode == 'abs':
+            return a > best + self.threshold
```

Comment:
Could these functions be private? BTW, it it necessary to make `in_cooldown` a property?

Reply:
Ok, it is a good idea.